### ISSUE TITLE
User Authentication Endpoints

### ISSUE THREAD

**Issue by borhara** [SHIV5T3R/SkillSync#6]
## Objective

The objective of this task is to implement a secure and efficient backend user authentication system that allows users to securely register, log in, and manage their accounts.

### Requirements

- [ ] **User Registration**

  Implement a user registration endpoint that allows new users to sign up for an account by providing a unique username, email, and password. Ensure that passwords are securely hashed and stored.

- [ ] **User Login**

  Develop a user login endpoint that allows registered users to authenticate themselves using their username/email and password combination. Upon successful authentication, generate and return a secure authentication token to be used for subsequent authenticated requests.

- [ ] **Token Management**

  Implement mechanisms for securely managing authentication tokens, including token generation, validation, and expiration. Tokens should expire after a certain period to ensure security.

- [ ] **Password Recovery**

  Implement a password recovery mechanism that allows users to reset their passwords in case they forget them. This may involve sending password reset links via email and securely verifying the user's identity before allowing them to reset their password.

- [ ] Security Measures

  Implement necessary security measures to prevent common vulnerabilities such as SQL injection, cross-site scripting (XSS), and cross-site request forgery (CSRF). Ensure that all user inputs are properly validated and sanitised.

- [ ] Logging and Monitoring

  Implement logging and monitoring functionalities to track authentication-related events and detect any suspicious activities.

- [ ] Testing

  Develop comprehensive unit tests and integration tests to ensure the reliability and security of the authentication system.

### Acceptance Criteria

- User authentication endpoints (registration, login, password recovery) are implemented and functioning correctly.
- Authentication tokens are securely generated, validated, and managed.
- The authentication system is secure and thoroughly tested.

**Comment by DynamiteBob17** [comment]
@jaybuurdd yeah, as far as logging is concerned, you can just annotate any class in which you want to log stuff with Lombok's @Slf4j annotation and you have a `log` instance at your disposal.

**Comment by DynamiteBob17** [comment]
I'll work on everything else, then.

**Comment by jaybuurdd** [comment]
> @jaybuurdd yeah, as far as logging is concerned, you can just annotate any class in which you want to log stuff with Lombok's `@Slf4j` annotation and you have a `log` instance at your disposal.

Okay cool I just saw how you did it in your code. Not much to do then lol 😂 I'll add logs to the entry point of the app and if there's anything I can take off your plate just let me know.

**Comment by jaybuurdd** [comment]
@DynamiteBob17 I'm going to setup env variables for the configs like the database.

**Comment by DynamiteBob17** [comment]
@jaybuurdd there is an example "env variables", i.e. the yaml config example file in the resources, the actual application.yaml config file used is gitignored for security purposes.

Also, I've whipped something up for user authentication with JWTs,

but if you want to help with

**Security Measures** (preventing XSS or CSRF attacks [SQL injection prevention is not necessary I think since so far no custom queries are used]),

**Password Recovery**

or **Testing** feel free to collaborate,

I haven't done much for these requirements.

Note: the testing part, if we want it to be thorough, would involve a whole lot of testing, and I don't have that kind of time 😢, as well as perhaps fully documenting the backend for the Swagger UI, so maybe you could work on this, or anyone else, if you can figure out what's going on haha.

I've pushed the changes I've made so far to a separate branch on my fork of this repo, so feel free to contribute to it since I don't really have a lot of time in general to work on stuff like this.

**Comment by DynamiteBob17** [comment]
And obviously, a lot can be logged throughout the whole application, as well 😃

**Comment by jaybuurdd** [comment]
> @jaybuurdd there is an example "env variables", i.e. the yaml config example file in the resources, the actual application.yaml config file used is gitignored for security purposes.
> 
> Also, I've whipped something up for user authentication with JWTs,
> 
> but if you want to help with
> 
> **Security Measures** (preventing XSS or CSRF attacks [SQL injection prevention is not necessary I think since so far no custom queries are used]),
> 
> **Password Recovery**
> 
> or **Testing** feel free to collaborate,
> 
> I haven't done much for these requirements.
> 
> Note: the testing part, if we want it to be thorough, would involve a whole lot of testing, and I don't have that kind of time 😅, as well as perhaps fully documenting the backend for the Swagger UI, so maybe you could work on this, or anyone else, if you can figure out what's going on haha.
> 
> I've pushed the changes I've made so far to a separate branch on my fork of this repo, so feel free to contribute to it since I don't really have a lot of time in general to work on stuff like this.

Nice I added my own application.yaml now for that.

I'll take care of password recovery and logging what you have in your branch.

**Comment by DynamiteBob17** [comment]
Ok, so the backend is secure enough for now as far as my fork is concerned haha, all we need is the password recovery system and perhaps writing some tests.

**Comment by jaybuurdd** [comment]
> Ok, so the backend is secure enough for now as far as my fork is concerned haha, all we need is the password recovery system and perhaps writing some tests.

Cool I should have it wrapped up by Wednesday, I'll be sure to pull your updates too. Also, started on some tests.

**Comment by jaybuurdd** [comment]
Wrapping it up today. Was too busy to get it done during the week, like I planned.

**Comment by jaybuurdd** [comment]
@DynamiteBob17 made a PR to your branch: https://github.com/DynamiteBob17/SkillSync/pull/1 didn't get to the unit tests, I'll do it later.

**Comment by Barungit** [comment]
Hey, anything where I can contribute in here? Password recovery done?

**Comment by jaybuurdd** [comment]
> Hey, anything where I can contribute in here? Password recovery done?

Already done. We made a PR covering the requirements for now. We do need unit testing though if you wanted to start on that.
